### ISSUE TITLE
Miscompilation of hash/crc64: return pointer is not restored on tail call

### ISSUE THREAD

**Issue by aykevl** [tinygo-org/tinygo#1467]
I have noticed a miscompilation in hash/crc64:

    tinygo test -opt=1 hash/crc64

I have tracked it to `hash/crc64.tableSum`, which looks like this:

```go
func tableSum(t *Table) uint64 {
	var a [2048]byte
	b := a[:0]
	if t != nil {
		for _, x := range t {
			b = appendUint64(b, x)
		}
	}
	return Checksum(b, MakeTable(ISO))
}
```

If I change it to this, the test passes:

```go
func tableSum(t *Table) uint64 {
	var a [2048]byte
	b := a[:0]
	if t != nil {
		for _, x := range t {
			b = appendUint64(b, x)
		}
	}
	checksum := Checksum(b, MakeTable(ISO))
	return checksum
}
```

The only difference are the last two lines, before it would return the result of `Checksum` directly and afterwards it returns them in a new variable. This _should_ not make a difference, except that the debug information is slightly different. In fact, when I compare the IR the only difference is two calls to `llvm.dbg.value` which are added in the working version.

I suspect the bug lies in the coroutine code as the output after the coroutine transformation is quite different, as a lot more extra calls are inserted in the working version (for example, a call to `(*internal/task.Task).getReturnPtr`). This tells me two things:

1. The coroutine optimization currently doesn't ignore calls to `llvm.dbg.value` while it should (debug information should not affect the program itself in any way).
2. It appears there is a bug in the coroutine tail call optimization.

To see the IR output, you can dump the function with a small change to the TinyGo source:

```diff
$ git diff
diff --git a/transform/optimizer.go b/transform/optimizer.go
index 917083ce..30b40c22 100644
--- a/transform/optimizer.go
+++ b/transform/optimizer.go
@@ -117,10 +117,12 @@ func Optimize(mod llvm.Module, config *compileopts.Config, optLevel, sizeLevel i
        switch config.Scheduler() {
        case "coroutines":
                // Lower async as coroutines.
+               mod.NamedFunction("hash/crc64.tableSum").Dump()
                err := LowerCoroutines(mod, config.NeedsStackObjects())
                if err != nil {
                        return []error{err}
                }
+               mod.NamedFunction("hash/crc64.tableSum").Dump()
        case "tasks":
                // No transformations necessary.
        case "none":
```

and run `tinygo test -opt=1 hash/crc64`.

@niaow can you maybe take a look?

**Comment by niaow** [comment]
I will try to look at this soon.

**Comment by niaow** [comment]
This turns out to be a very straightforward issue. The call to `MakeTable` clobbers the return buffer pointer. The return buffer is not saved, so the `Checksum` call writes into `MakeTable`'s return buffer instead of the return buffer that was provided by the caller of `tableSum`.

> The coroutine optimization currently doesn't ignore calls to llvm.dbg.value while it should (debug information should not affect the program itself in any way).

Right now, the code to detect a tail call is somewhat oversimplified in that it just looks for a call immediately preceding a return. Ideally we should be a bit smarter, but right now we are not.

**Comment by deadprogram** [comment]
This was released as part of v0.16.0 so now closing.
